src/main/java/com/statsig/androidsdk/Statsig.kt

```diff
@@ -23,431 +33,423 @@ interface IStatsigCallback {
-         * use updateUser()
-         */
-        @JvmOverloads
-        @JvmStatic
-        fun initializeAsync(
-            application: Application,
-            sdkKey: String,
-            user: StatsigUser? = null,
-            callback: IStatsigCallback? = null,
-            options: StatsigOptions = StatsigOptions(),
-        ) {
-            GlobalScope.async {
-                runBlocking {
-                    initialize(application, sdkKey, user, options)
-                }
-                if (callback != null) {
-                    callback.getStatsigHandler().post {
-                        callback.onStatsigInitialize()
-                    }
+object Statsig {
+
+    private const val SHARED_PREFERENCES_KEY: String = "com.statsig.androidsdk"
+
+    private const val INITIALIZE_RESPONSE_KEY: String = "Statsig.INITIALIZE_RESPONSE"
+
+    private val statsigJob = SupervisorJob()
+    private val statsigScope = CoroutineScope(statsigJob + Dispatchers.Main)
+
+    private val gson = Gson()
+
+    private var state: StatsigState? = null
+    private var user: StatsigUser? = null
+
+    private var pollingJob: Job? = null
+    private lateinit var application: Application
+    private lateinit var sdkKey: String
+    private lateinit var options: StatsigOptions
+    private lateinit var lifecycleListener: StatsigActivityLifecycleListener
+
+    private lateinit var logger: StatsigLogger
+    private lateinit var statsigMetadata: StatsigMetadata
+
+    @VisibleForTesting
+    internal var statsigNetwork: StatsigNetwork = StatsigNetwork()
+
+    /**
+     * Initializes the SDK for the given user.  Initialization is complete when the callback
+     * is invoked
+     * @param application - the Android application Statsig is operating in
+     * @param sdkKey - a client or test SDK Key from the Statsig console
+     * @param user - the user to associate with feature gate checks, config fetches, and logging
+     * @param callback - a callback to execute when initialization is complete
+     * @param options - advanced SDK setup
+     * Checking Gates/Configs before initialization calls back will return default values
+     * Logging Events before initialization will drop those events
+     * Susequent calls to initialize will be ignored.  To switch the user or update user values,
+     * use updateUser()
+     */
+    @JvmOverloads
+    @JvmStatic
+    fun initializeAsync(
+        application: Application,
+        sdkKey: String,
+        user: StatsigUser? = null,
+        callback: IStatsigCallback? = null,
+        options: StatsigOptions = StatsigOptions(),
+    ) {
+        statsigScope.launch {
+            initialize(application, sdkKey, user, options)
+            if (callback != null) {
+                // Run on the handler if we are already in the same looper. Else post on it via coroutines.
+                withContext(callback.getStatsigHandler().asCoroutineDispatcher().immediate) {
```

Comment:
The callback was for supporting asynchronous initialization in java, in a world before support for CompletableFuture (we still need to support android API levels below 24).

If there is a better way to accomplish this, I'm all ears

Reply:
This can be done with a simple callback. Documentation can point out that the callback would be run on the main thread. No need for a `CompletableFuture`.